README.md

```diff
@@ -35,54 +34,44 @@ This makes it much easier to contribute new packages packages, since there's no
 # Detailed design
 [design]: #detailed-design
 
-This RFC establishes the convention of `pkgs/unit/${toLower (substring 0 4 name)}/${name}` "unit" directories for the definitions of the Nix packages `pkgs.${name}` in nixpkgs.
-The `pkg-fun.nix` files in all unit directories are automatically discovered, called using `pkgs.callPackage` and added to the `pkgs` set.
+This RFC establishes the standard of using `pkgs/unit/${shard}/${name}` "unit" directories for the definitions of the Nix packages `pkgs.${name}` in nixpkgs, where `shard = toLower (substring 0 2 name)`.
+All unit directories are automatically discovered and incorporated into the `pkgs` set using `pkgs.${name} = pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix { }`.
 
-These requirements will be checked using CI:
-1. The `pkgs/unit` directory must only contain unit directories, and only in subdirectories of the form `${substring 0 4 name}/${name}`.
-2. <a id="req-ref-out"/> Files inside a unit directory must not reference files outside that unit directory.
-3. <a id="req-ref-in"/> Files outside a unit directory must not reference files inside a unit directory, except for definitions of attributes in `all-packages.nix` and the auto-calling logic.
-4. The definition of a package in the unit directory is the one `pkgs.<name>` points to.
-5. To avoid problems with merges, if a package attribute is defined by a unit directory, an attribute of the same name in `pkgs/top-level/all-packages.nix` (or `pkgs/top-level/aliases.nix`) must not redefine it in terms of something other than the unit.
+These requirements will be checked by CI:
+1. Structure: The `pkgs/unit` directory must only contain unit directories, and only in subdirectories of the form `${shard}/${name}`.
+  Each unit directory must contain at least a `pkg-fun.nix` file, but may contain arbitrary other files and directories.
+2. Semantics: Calling `pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix {}` must return a derivation that can be built directly with `nix-build`.
+3. Context-freedom: Files inside a unit directory must not reference files outside that unit directory, and the other way around.
+  The only exception is the next requirement.
+4. Custom arguments: If `pkgs/top-level/all-packages.nix` contains a definition for the attribute `${name}` and the unit directory `pkgs/unit/${shard}/${name}` exists, then the attribute value must be defined as `pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix args`, where `args` may be freely chosen.
 
-This convention is optional, but it will be applied to all existing packages where possible. Nixpkgs reviewers may encourage contributors to use this convention without enforcing it.
+This standard must be followed for newly added packages.
+A treewide migration to this standard will be performed for all existing packages that can satisfy the requirements.
```

Comment:
Does that include `args != { }`? If not, that's ok, but then:
```suggestion
A treewide migration to this standard will be performed for existing packages that can satisfy the requirements.
```

Reply:
I think the tooling should do all the migrations that it can, which then includes when `args != {}`, because if the tooling doesn't do it, users might do it anyways, it is something that's supported after all. It's not great how one can't know what arguments `pkg-fun.nix` files are called with anymore, but this is the case in current nixpkgs anyways, and migrating or not for `args != {}` won't change that.

I'm a bit confused by your suggestion though, if only `args == {}` would be migrated, then that would be an additional requirement for the migration only, but since I don't think we should do that, the sentence is good as is. It should be fine to remove the word `all` in both cases though, doesn't really change the meaning.

Reply:
If the tool can migrate `args != { }`, all the better.
I assumed that this sentence was about the automated part only. I do think they should eventually be migrated, whether automatically or by hand.

Reply:
> one can't know what arguments `pkg-fun.nix` files are called with anymore, but this is the case in current nixpkgs anyway

Correct, which is why we have open issue:
 - https://github.com/nixpkgs-architecture/issues/issues/8

Reply:
In the latest commit I removed the `all`, I also moved these sentences above the requirements.